core/controllers/acl_decorators.py

```diff
@@ -45,6 +47,72 @@
 (suggestion_models,) = models.Registry.import_models([models.NAMES.suggestion])
 
 
+def redirect_permanently(handler, redirection_url, expected_return_type):
+    """Redirects to the provided URL if the handler type is not JSON. The
+    status code used would be 301 for permanent redirection.
+
+    Args:
+        handler: function. The function to be decorated.
+        redirection_url: str. The URL to redirect to.
+        expected_return_type: str. The type of the response to be returned
+            in case of errors eg. html, json.
+
+    Raises:
+        PageNotFoundException: The page is not found.
+    """
+    if expected_return_type == feconf.HANDLER_TYPE_JSON:
+        raise handler.PageNotFoundException
+    else:
+        handler.redirect(redirection_url, permanent=True)
+
+
+def validate_and_redirect_based_on_classroom_name(
+        handler, topic_id, classroom_name_url_fragment,
+        url_substring):
+    """Validates classroom name and redirects to the URL with the
+    valid classroom name. If the classroom name provided is valid,
+    and no redirection is required, return True. Otherwise, return False.
+
+    Args:
+        handler: function. The function to be decorated.
+        topic_id: str. The id of the topic.
+        classroom_name_url_fragment: str. The classroom name provided
+            in the URL.
+        url_substring: str. The URL substring related to the entity.
+
+    Returns:
+        bool. Whether classroom name provided was valid or not.
+    """
+    lowercase_classroom_name = (
+        python_utils.url_unquote_plus(
+            classroom_name_url_fragment).replace('-', ' '))
+    topic_ids_for_classroom_pages = (
+        config_domain.TOPIC_IDS_FOR_CLASSROOM_PAGES.value)
+    topic_is_assigned_classroom = False
+    for classroom_dict in topic_ids_for_classroom_pages:
+        if topic_id in classroom_dict['topic_ids']:
+            if (
+                    lowercase_classroom_name ==
+                    classroom_dict['name'].lower()):
+                topic_is_assigned_classroom = True
+                break
+            else:
+                redirect_permanently(
+                    handler, '/learn/%s/%s' % (
+                        classroom_dict['name'].lower().replace(' ', '-'),
```

Comment:
Same thing, don't do conversions. Store the data directly and only check equality. I think we will want/need the freedom to define arbitrary url fragments and they may not match the classroom name exactly.